api/lib/firewall_functions.pl

```diff
@@ -52,7 +52,21 @@ sub read_addresses
     chomp $cidr;
     $cidr =~ /^(.*)\/(.*)$/;
     my $ipaddr = $1;
-    my $gw = `ip -o route list dev $interface  | tail -n 1 | awk '{print \$1}'`;
+    my $ndb = esmith::NetworksDB->open_ro();
+    my %props = $ndb->get($interface)->props;
+    my $gw;
+
+    if ((exists($props{'bootproto'}) && $props{'bootproto'} eq 'dhcp') || $interface eq 'ppp0' ) {
+        if ($props{'bootproto'} eq 'dhcp') {
+            # DHCP
+            $gw = `/sbin/ip -4 route show dev $interface default | /usr/bin/awk '{print \$3}'`;
+        } else {
+            # PPPoE
+            $gw = `/sbin/ip -4 route show dev ppp0 | /usr/bin/grep src | /usr/bin/awk '{print \$1}'`;
+        }
+    } else {
```

Comment:
Something like:
```suggestion
    if ($interface eq 'ppp0' ) {
            $gw = `/sbin/ip -4 route show dev ppp0 | /usr/bin/grep src | /usr/bin/awk '{print \$1}'`;
        } else if (exists($props{'bootproto'}) && $props{'bootproto'} eq 'dhcp')){
            $gw = `/sbin/ip -4 route show dev $interface default | /usr/bin/awk '{print \$3}'`;
         }
    } else {
```

Please note that my identation is note good.